app/services.go

```diff
@@ -0,0 +1,68 @@
+package app
+
+import (
+	"context"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/juju/errors"
+
+	"github.com/projecteru2/barrel/service"
+)
+
+func startServices(sigs <-chan os.Signal, disposeTimeout time.Duration, ss []service.Service) error {
+	chErr := make(chan error)
+	defer close(chErr)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	wg := sync.WaitGroup{}
+	wg.Add(len(ss))
+
+	for _, elm := range ss {
+		go func(serv service.Service) {
+			chErr <- startService(ctx, serv, disposeTimeout)
+			wg.Done()
+		}(elm)
+	}
+
+	select {
+	case <-sigs:
+		cancel()
+		wg.Wait()
+		return nil
+	case err := <-chErr:
+		wg.Wait()
+		cancel()
+		return err
+	}
+}
+
+func startServiceWithoutDisposeTimeout(ctx context.Context, serv service.Service) error {
+	disposable, serveErr := serv.Serve(ctx)
+	if err := disposable.Dispose(context.Background()); err != nil {
+		if serveErr != nil {
+			return errors.Wrap(serveErr, err)
+		}
+		return err
+	}
+	return serveErr
+}
+
+func startService(ctx context.Context, serv service.Service, disposeTimeout time.Duration) error {
+	if disposeTimeout == time.Duration(0) {
+		return startServiceWithoutDisposeTimeout(ctx, serv)
+	}
+
+	timeoutCtx, timeoutCancel := context.WithTimeout(context.Background(), disposeTimeout)
+	defer timeoutCancel()
+
+	disposable, serveErr := serv.Serve(ctx)
+	if err := disposable.Dispose(timeoutCtx); err != nil {
```

Comment:
followings are totally the same as above in startServiceWithoutDisposeTimeout func, shall we refining them as a func alone.

Reply:
ok, i didn't refining them because golang lint give me a warning that context should be first parameter of function but we need two context as parameter

Reply:
> ok, i didn't refining them because golang lint give me a warning that context should be first parameter of function but we need two context as parameter

now i make a new struct to refine the proc